### PR TITLE
Add Parlia consensus engine for Binance Smart Chain support

### PR DESCRIPTION
**Pull request by brendan-kelly**

### CODE REVIEW

**Comment by AlexeyAkhunov** [comment]
thank you! Can you please fix formatting in couple of files, you can see which in the "Files Changed" tab

File: core/types/transaction_signing.go [hunk 1]
```diff
@@ -277,6 +277,22 @@ func (sg Signer) SignatureValues(tx Transaction, sig []byte) (R, S, V *uint256.I
 	return R, S, V, nil
 }
 
+// EIP155Signer implements Signer using the EIP-155 rules. This accepts transactions which
+// are replay-protected as well as unprotected homestead transactions.
+type EIP155Signer struct {
```
Comment (AlexeyAkhunov) [inline core/types/transaction_signing.go:282]: You do not need to re-introduce EIP155Signer, because it has been factored into the more generic signer type. If you want specifically EIP155 signer, you can instantiate signer with:
```
		signer.protected = true
		signer.chainID.Set(&chainId)
		signer.chainIDMul.Mul(&chainId, u256.Num2)
```

then you don't need to re-introduce EIP155Signer type

File: eth/backend.go [hunk 2]
```diff
@@ -89,6 +90,8 @@ type Ethereum struct {
 
 	engine consensus.Engine
 
+	APIBackend *EthAPIBackend
```
Comment (AlexeyAkhunov) [inline eth/backend.go:93]: We do not want to reintroduce EthAPIBackend here. All RPC communication is happening via RPC daemon

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -32,6 +32,21 @@ import (
 	"github.com/ledgerwatch/log/v3"
 )
 
+// Covalent change: Empty Backend interface used for PublicBlockChainAPI below
```
Comment (AlexeyAkhunov) [inline internal/ethapi/api.go:35]: Please do not bring this back, all RPC comms need to go via RPC Daemon. We are also trying to gradually retire Backend approach. This package is currently only used to contain some types and functions that are used in RPC Daemon, and needs some cleanup

File: params/config.go [hunk 4]
```diff
@@ -85,6 +85,7 @@ var (
 		PetersburgBlock:     big.NewInt(7_280_000),
 		IstanbulBlock:       big.NewInt(9_069_000),
 		MuirGlacierBlock:    big.NewInt(9_200_000),
+		RamanujanBlock:      big.NewInt(0),
```
Comment (AlexeyAkhunov) [inline params/config.go:88]: This should probably left as `nil` and not zero (which would mean the this fork is activated straight away)

File: params/config.go [hunk 5]
```diff
@@ -108,6 +109,7 @@ var (
 		PetersburgBlock:     big.NewInt(4_939_394),
 		IstanbulBlock:       big.NewInt(6_485_846),
 		MuirGlacierBlock:    big.NewInt(7_117_117),
+		RamanujanBlock:      big.NewInt(0),
```
Comment (AlexeyAkhunov) [inline params/config.go:112]: This should probably left as nil and not zero (which would mean the this fork is activated straight away)

File: params/config.go [hunk 6]
```diff
@@ -128,6 +130,7 @@ var (
 		ByzantiumBlock:      big.NewInt(1_035_301),
 		ConstantinopleBlock: big.NewInt(3_660_663),
 		PetersburgBlock:     big.NewInt(4_321_234),
+		RamanujanBlock:      big.NewInt(0),
```
Comment (AlexeyAkhunov) [inline params/config.go:133]: This should probably left as nil and not zero (which would mean the this fork is activated straight away)

File: params/config.go [hunk 7]
```diff
@@ -463,10 +482,12 @@ func (c *ChainConfig) String() string {
 		engine = c.Ethash
 	case c.Clique != nil:
 		engine = c.Clique
+	case c.Parlia != nil:
+		engine = c.Parlia
 	default:
 		engine = "unknown"
 	}
-	return fmt.Sprintf("{ChainID: %v Homestead: %v DAO: %v DAOSupport: %v EIP150: %v EIP155: %v EIP158: %v Byzantium: %v Constantinople: %v Petersburg: %v Istanbul: %v, Muir Glacier: %v, Berlin: %v, London: %v, Arrow Glacier: %v, Engine: %v}",
+	return fmt.Sprintf("{ChainID: %v Homestead: %v DAO: %v DAOSupport: %v EIP150: %v EIP155: %v EIP158: %v Byzantium: %v Constantinople: %v Petersburg: %v Istanbul: %v, Muir Glacier: %v, Ramanujan: %v, Berlin: %v, London: %v, Arrow Glacier: %v, Engine: %v}",
```
Comment (AlexeyAkhunov) [inline params/config.go:490]: Can this be re-worked so that it only prints the fork with non nil block number? So we do not display irrelevant fork numbers on main net and BSC?

File: params/config.go [hunk 8]
```diff
@@ -544,11 +566,21 @@ func (c *ChainConfig) IsByzantium(num uint64) bool {
...
+}
+
 // IsConstantinople returns whether num is either equal to the Constantinople fork block or greater.
 func (c *ChainConfig) IsConstantinople(num uint64) bool {
 	return isForked(c.ConstantinopleBlock, num)
 }
 
+// IsRamanujan returns whether num is either equal to the IsRamanujan fork block or greater.
+func (c *ChainConfig) IsRamanujanBigInt(num *big.Int) bool {
+	return configNumEqual(c.RamanujanBlock, num)
```
Comment (AlexeyAkhunov) [inline params/config.go:581]: This should be using `isForked` function instead of `configNumEqual` ?

File: consensus/consensus.go [hunk 9]
```diff
@@ -104,15 +108,15 @@ type Engine interface {
 	//
 	// Note: The block header and state database might be updated to reflect any
 	// consensus rules that happen at finalization (e.g. block rewards).
-	Finalize(config *params.ChainConfig, header *types.Header, state *state.IntraBlockState, txs []types.Transaction, uncles []*types.Header, r types.Receipts, e EpochReader, chain ChainHeaderReader, syscall SystemCall) error
+	Finalize(config *params.ChainConfig, header *types.Header, state *state.IntraBlockState, txs []types.Transaction, uncles []*types.Header, r types.Receipts, systemTxs *[]*types.Transaction, usedGas *uint64, e EpochReader, chain ChainHeaderReader, syscall SystemCall) error
```
Comment (AlexeyAkhunov) [inline consensus/consensus.go:111]: Since `types.Transaction` is an interface in Erigon, unlike concrete type in go-ethereum, it should not be used
as `*[]*types.Transaction`, but instead `*[]types.Transaction`. Please change it accordingly in other places. Otherwise it is pointer to the interface, which is unnecessary overhead and less readable.

Also, instead of receiving pointer to the slice `*[]`, so that `Finalize` can modify it, it can return the slice as one of the return values. The only scenarios where I'd think passing pointer to a slice would help is if the slice is expensive to create and needs to be re-used, but I don't think it is the case here.

To summarise, the signature should be:
```
Finalize(config *params.ChainConfig, header *types.Header, state *state.IntraBlockState, txs []types.Transaction, uncles []*types.Header, r types.Receipts, e EpochReader, chain ChainHeaderReader, syscall SystemCall) (systemTxs []types.Transaction, usedGas uint64, err error)
```

then it is clear that `Finalize` is suppose to produce and return a slice of system transaction and also amount of gas used for those (I assume)

File: consensus/consensus.go [hunk 10]
```diff
@@ -144,3 +148,12 @@ type PoW interface {
 	// Hashrate returns the current mining hashrate of a PoW consensus engine.
 	Hashrate() float64
 }
+
+type PoSA interface {
```
Comment (AlexeyAkhunov) [inline consensus/consensus.go:152]: What does `PoSA` stand for? Can you add a comment

**Comment by AlexeyAkhunov** [comment]
Can you also remove the empty branches and ineffectual assignments (can also be found on "Files Changed" tab.

Please do not try to introduce JSON RPC API inside Erigon, and leave it for later. If there are bsc specific APIs that needed to the added to the RPC Daemon, then we can figure out necessary grpc interfaces between RPC Daemon and Erigon. Currently the interface definition is here: https://github.com/ledgerwatch/interfaces/blob/master/remote/ethbackend.proto
For example, we are adding special functions for the merge (Proof of Stake)

File: eth/backend.go (hunk 2 above)
Comment (brendan-kelly) [reply eth/backend.go:93]: What should I use an alternative programmatically?

File: internal/ethapi/api.go (hunk 3 above)
Comment (brendan-kelly) [reply internal/ethapi/api.go:35]: What should I use an alternative programmatically?

File: params/config.go (hunk 7 above)
Comment (brendan-kelly) [reply params/config.go:490]: Should I just leave this out then until the other changes for the other PRs (CLI and rest of BSC changes) are ready?

File: params/config.go (hunk 8 above)
Comment (brendan-kelly) [reply params/config.go:581]: I will double check with what BSC did here

Comment (brendan-kelly) [reply params/config.go:581]: I use `configNumEqual` here since the definitions are:
```
func isForked(s *big.Int, head uint64) bool
func configNumEqual(x, y *big.Int) bool
```

There is a compilation error when using `isForked`:
```
params/config.go:581:19: cannot use num (type *big.Int) as type uint64 in argument to isForked
```

File: eth/backend.go (hunk 2 above)
Comment (AlexeyAkhunov) [reply eth/backend.go:93]: At the moment, the code goes into `cmd/rpcdaemon/commands`, but I would leave it for a separate PR. We need to see what extra RPC requests need to be implemented, and how necessarily data can be delivered to the RPC Daemon. Due to architecture of separability of RPC Daemon and Erigon, there needs to be a way of delivering data both local and remotely

File: internal/ethapi/api.go (hunk 3 above)
Comment (AlexeyAkhunov) [reply internal/ethapi/api.go:35]: The code for extra RPC currently goes into `cmd/rpcdaemon/commands`, as I mentioned in one of the previous comment

File: params/config.go (hunk 7 above)
Comment (AlexeyAkhunov) [reply params/config.go:490]: Yes, you can leave config changes out, and bring it in separately, with re-work of this printout

File: params/config.go (hunk 8 above)
Comment (AlexeyAkhunov) [reply params/config.go:581]: Perhaps in BSC, this hardfork is different, and instead of changing the rules after certain block, it has to execute something on that block, then I would understand the semantics. Can you confirm what Ramanujan hardfork is doing

Comment (brendan-kelly) [reply params/config.go:580]: Here is what BSC is doing - https://github.com/binance-chain/bsc/blob/master/params/config.go#L732 and here is what is implemented in Erigon - https://github.com/covalenthq/erigon/blob/devel/params/config.go#L710. What I've added is basically a workaround to deal with the 2 forks using `big.Int` vs `uint64`.

This is all so I can do the checks in these locations: https://github.com/ledgerwatch/erigon/blob/8791d432c846c118e64239ba26e9e8d6416ef92c/consensus/parlia/ramanujanfork.go#L18 https://github.com/ledgerwatch/erigon/blob/8791d432c846c118e64239ba26e9e8d6416ef92c/consensus/parlia/ramanujanfork.go#L31 https://github.com/ledgerwatch/erigon/blob/8791d432c846c118e64239ba26e9e8d6416ef92c/consensus/parlia/ramanujanfork.go#L38

File: core/types/transaction_signing.go [hunk 11]
```diff
@@ -347,3 +347,17 @@ func DeriveChainId(v *uint256.Int) *uint256.Int {
 	v = new(uint256.Int).Sub(v, u256.Num35)
 	return v.Div(v, u256.Num2)
 }
+
+// Hash returns the hash to be signed by the sender.
+// It does not uniquely identify the transaction.
+func (sg Signer) Hash(tx Transaction) common.Hash {
```
Comment (AlexeyAkhunov) [inline core/types/transaction_signing.go:353]: can you take a look at `SigningHash` function of type `LegacyTx` in the file `core/legacy_tx.go`. If you only need this function for legacy-type txs, you can try to use existing function?

File: consensus/parlia/parlia.go [hunk 12]
```diff
@@ -0,0 +1,1278 @@
...
+
+	expectedTx = func(initExpectedTx interface{}) types.Transaction {
+		expectedTx, ok := initExpectedTx.(types.Transaction)
+		if !ok {
+			panic("Oops")
+		}
+		return expectedTx
+	}(initExpectedTx)
+
+	expectedHash := p.signer.Hash(expectedTx)
```
Comment (AlexeyAkhunov) [inline consensus/parlia/parlia.go:1131]: Can it be `expectedTx.SigningHash(p.chainConfig.ChainID)` ?

File: core/state_transition.go [hunk 13]
```diff
@@ -402,6 +402,7 @@ func (st *StateTransition) TransitionDb(refunds bool, gasBailout bool) (*Executi
 	if st.evm.ChainRules().IsLondon {
 		effectiveTip = cmath.Min256(st.tip, new(uint256.Int).Sub(st.gasFeeCap, st.evm.Context().BaseFee))
 	}
+	// consensus engine is parlia
```
Comment (AlexeyAkhunov) [inline core/state_transition.go:405]: If the code got removed from there, perhaps the comment has to be as well

**Review by AlexeyAkhunov — APPROVED** [review]
